File: bindings/python/examples/how_tos/outputs/features.py
import json
from dataclasses import asdict

from dotenv import load_dotenv
from iota_sdk import (
    AddressUnlockCondition,
    Client,
    Utils,
    SenderFeature,
    IssuerFeature,
    MetadataFeature,
    TagFeature,
    utf8_to_hex,
)

load_dotenv()

client = Client()

address = Utils.parse_bech32_address(
    'rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy')

address_unlock_condition = AddressUnlockCondition(address)

# Output with sender feature
nft_output = client.build_nft_output(
    nft_id='0x0000000000000000000000000000000000000000000000000000000000000000',
    unlock_conditions=[
        address_unlock_condition
    ],
    features=[
        SenderFeature(address)
    ],
)
outputs = [nft_output]

# Output with issuer feature
nft_output = client.build_nft_output(
    nft_id='0x0000000000000000000000000000000000000000000000000000000000000000',
    unlock_conditions=[
        address_unlock_condition,
    ],
    immutable_features=[
        IssuerFeature(address)
    ],
)
outputs.append(nft_output)

# Output with metadata feature
nft_output = client.build_nft_output(
    nft_id='0x0000000000000000000000000000000000000000000000000000000000000000',
    unlock_conditions=[
        address_unlock_condition,
    ],
    features=[
        MetadataFeature({'data': utf8_to_hex('Hello, World!')})
    ],
)
outputs.append(nft_output)

# Output with immutable metadata feature
nft_output = client.build_nft_output(
    nft_id='0x0000000000000000000000000000000000000000000000000000000000000000',
    unlock_conditions=[
        address_unlock_condition,
    ],
    immutable_features=[
        MetadataFeature({'data': utf8_to_hex('Hello, World!')})
    ],
)
outputs.append(nft_output)

# Output with tag feature
nft_output = client.build_nft_output(
    nft_id='0x0000000000000000000000000000000000000000000000000000000000000000',
    unlock_conditions=[
        address_unlock_condition
    ],
    features=[
        TagFeature(utf8_to_hex('Hello, World!'))
    ],
)
outputs.append(nft_output)

print(json.dumps([asdict(o) for o in outputs], indent=2))

File: bindings/nodejs/examples/how_tos/outputs/features.ts
// Copyright 2021-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    Client,
    initLogger,
    Utils,
    UnlockCondition,
    AddressUnlockCondition,
    MetadataFeature,
    SenderFeature,
    TagFeature,
    IssuerFeature,
    utf8ToHex,
} from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./how_tos/outputs/features.ts

// Build outputs with all features
async function run() {
    initLogger();

    const client = await Client.create({});

    try {
        const ed25519Address = Utils.parseBech32Address(
            'rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy',
        );

        const addressUnlockCondition: UnlockCondition =
            new AddressUnlockCondition(ed25519Address);

        // Output with sender feature
        const nftOutputWithSender = await client.buildNftOutput({
            nftId: '0x0000000000000000000000000000000000000000000000000000000000000000',
            unlockConditions: [addressUnlockCondition],
            features: [new SenderFeature(ed25519Address)],
        });

        // Output with issuer feature
        const nftOutputWithIssuer = await client.buildNftOutput({
            nftId: '0x0000000000000000000000000000000000000000000000000000000000000000',
            unlockConditions: [addressUnlockCondition],
            immutableFeatures: [new IssuerFeature(ed25519Address)],
        });

        // Output with metadata feature
        const nftOutputWithMetadata = await client.buildNftOutput({
            nftId: '0x0000000000000000000000000000000000000000000000000000000000000000',
            unlockConditions: [addressUnlockCondition],
            features: [
                new MetadataFeature({ data: utf8ToHex('Hello, World!') }),
            ],
        });

        // Output with immutable metadata feature
        const nftOutputWithImmutableMetadata = await client.buildNftOutput({
            nftId: '0x0000000000000000000000000000000000000000000000000000000000000000',
            unlockConditions: [addressUnlockCondition],
            immutableFeatures: [
                new MetadataFeature({ data: utf8ToHex('Hello, World!') }),
            ],
        });

        // Output with tag feature
        const nftOutputWithTag = await client.buildNftOutput({
            nftId: '0x0000000000000000000000000000000000000000000000000000000000000000',
            unlockConditions: [addressUnlockCondition],
            features: [new TagFeature(utf8ToHex('Hello, World!'))],
        });

        console.log(
            JSON.stringify(
                [
                    nftOutputWithSender,
                    nftOutputWithIssuer,
                    nftOutputWithMetadata,
                    nftOutputWithImmutableMetadata,
                    nftOutputWithTag,
                ],
                null,
                2,
            ),
        );
    } catch (error) {
        console.error('Error: ', error);
    }
}

void run().then(() => process.exit());
